### PR TITLE
Improve weight documentation

### PR DESCRIPTION
**Pull request by athei**
Add documentation for the following:

* What is the meaning of one unit of weight
* Why abstract over database access costs?
* Post dispatch weight correction.
* How to use the weight annotation.

I am not a native speaker. So adjust your expectations accordingly and please feel free to correct my writing and edit this PR.

### CODE REVIEW

File: current/learn-substrate/weight.md [hunk 1]
```diff
@@ -19,6 +19,13 @@ used in mandatory inherent extrinsics) will generally be limited through economi
 simple terms, through transaction fees. The fee implications of the weight system are covered in the
 [Transaction Fees document](../runtime/fees).
 
+Substrates defines one unit of weight as one picosecond of execution time on some reference machine.
+Currently, that is an Intel Core i7-7700K CPU with 64GB of RAM and a NVMe SSD. Which reference
+machine is used ultimately depends on what are the hardware requirements the runtime developer
+wants to put on the validators. Weights are determined by benchmarking every dispatchable on this
+reference machine. Over time, the weight values will be adjusted to keep up with hardware
+and software improvements.
+
```
Comment (JoshOrndorff) [inline current/learn-substrate/weight.md]: I thought the reference hardware was supposed to be a universal standard. Having auniversal standard solves the problem of a pallet being written initially for one runtime, but then used in other runtimes. If different runtimes can have different reference hardware, then they will calculate weights differently. And we're straight back to pallets not being reusable in multiple runtimes.

My understanding was that if my chain wants to allow weaker hardware than the reference hardware you stated, then I should adjust my target block time and max block weight, not the definition of the reference hardware.

cc @shawntabrizi

File: current/learn-substrate/weight.md [hunk 2]
```diff
@@ -53,13 +60,38 @@ are available:
...
   operation should gracefully error.
 
+One factor that contributes greatly to the execution time of a dispatchable is the amount of
+database accesses that are performed by that dispatchable. At the same time the cost of a database
+access is greatly depended on the employed database backend and storage hardware. For that reason
+the weight calculations are parameterized over the weight costs of database reads and writes. These
+costs are determined by benchmarking each available database backend on some reference hardware.
+This allows switching the database backend without changing all weight calculations.
+
+In addition to only using constants for the ahead of dispatch weight calculation the developer
```
Comment (JoshOrndorff) [inline current/learn-substrate/weight.md]:
```suggestion
In addition to only using constants for the ahead-of-dispatch weight calculation the developer
```

File: current/learn-substrate/weight.md [hunk 3]
```diff
@@ -53,13 +60,38 @@ are available:
...
+has the ability to factor in the input parameters of the given dispatchable. This can be useful
+when the execution time depends on for example the length of one parameter. It is important that
+these calculations do not entail any meaningful work itself. The ahead of dispatch maximum weight
```
Comment (JoshOrndorff) [inline current/learn-substrate/weight.md]:
```suggestion
these calculations do not entail any meaningful work itself. The ahead-of-dispatch maximum weight
```

File: current/learn-substrate/weight.md [hunk 4]
```diff
@@ -53,13 +60,38 @@ are available:
...
+## Post Dispatch Weight Correction
+
+There are cases where the actual weight of a dispatchable is not trivially computable from it's inputs
+and is therefore only known after the dispatchable is executed. Without any means of correcting the
+weight after the dispatch we would be constantly over estimating and subsequently overcharging
+for those dispatchables as we must assume the worst case ahead of dispatch for the chain to be safe.
+
+The post dispatch weight correction allows any dispatchable to return it's _actual weight_ after
+it was executed. This weight must be less or equal than the ahead of dispatch worst case weight. For
+a user to be allowed to include a extrinsic they still must be able to pay for the ahead of
```
Comment (JoshOrndorff) [inline current/learn-substrate/weight.md]:
```suggestion
a user to be allowed to include an extrinsic they still must be able to pay for the ahead-of-
```

File: current/runtime/fees.md [hunk 5]
```diff
@@ -109,30 +109,86 @@ Some transactions warrant limiting resources with other strategies. For example:
...
+```
 
-### Normal Dispatches
+This dispatchable does one database read and two database writes in addition to other things
+that add the rest that is 20_000 weight. A database access is general every time when a value
+is accessed that is declared inside the `decl_storage!` block. However, only unique accesses are
+counted as once a value is accessed it is cached and accessing it again does not result in a
+database operation.
+
+Multiple reads of the same value count as one read.Multiple writes of the same value are only
```
Comment (JoshOrndorff) [inline current/runtime/fees.md]:
```suggestion
Multiple reads of the same value count as one read. Multiple writes of the same value are only
```

File: current/runtime/fees.md [hunk 6]
```diff
@@ -146,15 +202,53 @@ validators to abuse these types of dispatches, they may not be included in block
...
 ### Custom Weights
 
-Implementing a custom weight calculation function can vary in complexity.
-
-A weight calculation function must provide two trait implementations:
+Instead of using the default weight annotations described above one can create a custom weight
+calculation type. This type must implement the follow traits:
 
 - [`WeighData<T>`]: To determine the weight of the dispatch.
 - [`ClassifyDispatch<T>`]: To determine the class of the dispatch.
```
Comment (JoshOrndorff) [inline current/runtime/fees.md:254]: I think it also needs to implement PaysFees

File: current/learn-substrate/weight.md (hunk 1 above)
Comment (shawntabrizi) [reply current/learn-substrate/weight.md]: @JoshOrndorff you are correct here as far as I understand.

If your computer is 2x slower than our reference machine, you would set the `MaximumBlockWeight` to 1/2 the amount or you would double your block time.

In general though, FRAME makes an assumption about the hardware which defines our base units.

Comment (shawntabrizi) [reply current/learn-substrate/weight.md]: Maybe I dont understand the misunderstanding :)

Comment (athei) [reply current/learn-substrate/weight.md]: Okay that makes sense. I will change this to clearly state that the default hardware is fixed and that developers should tweak the `MaximumBlockWeight`.

File: current/runtime/fees.md (hunk 6 above)
Comment (athei) [reply current/runtime/fees.md:254]: That is correct. I will add it.

Comment (athei) [reply current/runtime/fees.md:254]: Fixed.

File: current/learn-substrate/weight.md (hunk 1 above)
Comment (athei) [reply current/learn-substrate/weight.md]: Fixed.

**Review by shawntabrizi — APPROVED** [review]
looks great

**Review by joepetrowski — APPROVED** [review]
Thanks for this!

**Pull request merged**